File: services/py/quant_core/domain/pure/compute_metrics.py
"""Pure projector: ``(StockMeta, kline_bars) -> StockMetrics``.

Persisted to ``data/stock_metas.parquet`` after each kline sync so the
list-view payload doesn't recompute on every read (see
``docs/perf/storage-unify-rollout.md`` item 9). Trigger: NestJS
``KlineWorker.process`` calls ``upsert_stock_metrics_for_code`` once
the new bars are written, then this projector picks them up locally.

口径 single-source-of-truth still lives in
:mod:`quant_core.domain.pure.derive_metrics` for the price-derived
half; this module only adds the return windows (``ret_*``) on top.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from decimal import Decimal
from typing import TYPE_CHECKING, Final

from quant_core.domain.pure.derive_metrics import derive_metrics

if TYPE_CHECKING:
    from collections.abc import Sequence

    from quant_core.domain.types.kline import DailyBar
    from quant_core.domain.types.stock import StockMeta


# Trading-bar lookback per return window. ``1`` means "vs the bar
# immediately before ``latest``" — the convention the snapshot handler
# has been using since v1.
_RETURN_WINDOWS: Final[tuple[tuple[str, int], ...]] = (
    ("ret_1d", 1),
    ("ret_5d", 5),
    ("ret_10d", 10),
    ("ret_20d", 20),
    ("ret_90d", 90),
    ("ret_250d", 250),
)


@dataclass(frozen=True, slots=True)
class StockMetrics:
    """Persisted projection of the snapshot derived/returns block.

    Every field is nullable — a bar shortage on a young listing, a missing
    financial input, or a non-positive denominator all propagate to
    ``None`` so the UI's "—" path stays correct.

    ``asof`` is the latest kline ``trade_date`` the metrics were computed
    against. ``price`` (``close_qfq`` at ``asof``) rides along so the
    snapshot handler can serve the full row from this block alone,
    skipping a second kline read on every list request.
    """

    code: str
    asof: date | None
    price: Decimal | None
    # returns
    ret_1d: Decimal | None
    ret_5d: Decimal | None
    ret_10d: Decimal | None
    ret_20d: Decimal | None
    ret_90d: Decimal | None
    ret_250d: Decimal | None
    # derived (mirror of DerivedMetrics)
    mkt_cap: Decimal | None
    float_mkt_cap: Decimal | None
    pe_ttm: Decimal | None
    pe_dynamic: Decimal | None
    pb: Decimal | None
    peg: Decimal | None
    gross_margin_ttm: Decimal | None


_EMPTY_RETURNS: Final[dict[str, Decimal | None]] = {name: None for name, _ in _RETURN_WINDOWS}


def compute_metrics(meta: StockMeta, bars: Sequence[DailyBar]) -> StockMetrics:
    """Project a fresh :class:`StockMetrics` row for ``meta``.

    Args:
        meta: Current meta snapshot. Used for ``derive_metrics``
            (financials + share counts) and to thread the code through.
        bars: Trailing kline bars in ascending date order. The latest
            bar's ``close_qfq`` feeds both ``derive_metrics`` and the
            return-window math. May be empty — in that case every metric
            (including derived) is ``None`` and ``asof`` is ``None``.

    Returns:
        Frozen :class:`StockMetrics` ready for
        ``ParquetStockMetaRepo.upsert_metrics``.
    """
    if not bars:
        return StockMetrics(
            code=meta.code,
            asof=None,
            price=None,
            **_EMPTY_RETURNS,
            mkt_cap=None,
            float_mkt_cap=None,
            pe_ttm=None,
            pe_dynamic=None,
            pb=None,
            peg=None,
            gross_margin_ttm=None,
        )
    latest = bars[-1]
    latest_close = latest.close_qfq
    derived = derive_metrics(meta, latest_close)
    returns = _compute_returns(bars, latest_close)
    return StockMetrics(
        code=meta.code,
        asof=latest.trade_date,
        price=latest_close if latest_close > 0 else None,
        **returns,
        mkt_cap=derived.mkt_cap,
        float_mkt_cap=derived.float_mkt_cap,
        pe_ttm=derived.pe_ttm,
        pe_dynamic=derived.pe_dynamic,
        pb=derived.pb,
        peg=derived.peg,
        gross_margin_ttm=derived.gross_margin_ttm,
    )


def _compute_returns(
    bars: Sequence[DailyBar], latest_close: Decimal
) -> dict[str, Decimal | None]:
    """Fractional change of ``latest_close`` vs the close N bars back.

    Returns ``None`` for any window where the bar count is insufficient
    or the historical close is non-positive.
    """
    if latest_close <= 0:
        return dict(_EMPTY_RETURNS)
    out: dict[str, Decimal | None] = {}
    for name, window in _RETURN_WINDOWS:
        if len(bars) <= window:
            out[name] = None
            continue
        base_close = bars[-1 - window].close_qfq
        if base_close <= 0:
            out[name] = None
            continue
        out[name] = (latest_close - base_close) / base_close
    return out

File: apps/api/src/modules/stock-meta/domain/pure/compute-metrics.ts
/**
 * Pure projector from `(StockMetaDto, BarLike[])` into the persisted
 * snapshot row that {@link LocalStockMetaWriterService.upsertMetrics}
 * writes to `data/stock_metas.parquet`.
 *
 * - `bars` is ascending by trade_date; `bars[-1]` is the latest.
 * - Empty `bars` yields a row with every metric null.
 * - Non-positive latest close → returns are null; the derived block
 *   still routes through `deriveMetrics` (which short-circuits to
 *   EMPTY when `price <= 0`).
 * - `wcmi` is computed by the dedicated scoring engine in
 *   `wcmi-scoring.ts`; see its docstring for the full formula.
 */

import type { StockMetaDto } from '@quant/shared';

import { D, type Dec } from './decimal-config.js';
import { deriveMetrics } from './derive-metrics.js';

export interface BarLike {
  readonly trade_date: string; // ISO YYYY-MM-DD
  readonly open_qfq: number;
  readonly high_qfq: number;
  readonly low_qfq: number;
  readonly close_qfq: number;
  /** Raw share volume (10k 股 for A-share). Used by the FOMO
   *  soft-penalty turnover guard — `0` is a safe default for fixtures
   *  that don't exercise that branch. */
  readonly volume: number;
  /** Daily turnover in 元 (akshare convention). Same guard as above. */
  readonly turnover: number;
  /** Pre-computed MAs from `KlineBar` (qfq close). `null` when the
   *  trailing window is shorter than the MA period. Consumed by the
   *  WCMI sub-score engine (see `wcmi-subscores/`). */
  readonly ma5: number | null;
  readonly ma10: number | null;
  readonly ma20: number | null;
  readonly ma60: number | null;
}

export interface StockMetrics {
  readonly code: string;
  readonly asof: string | null;
  readonly price: Dec | null;
  readonly ret_1d: Dec | null;
  readonly ret_5d: Dec | null;
  readonly ret_10d: Dec | null;
  readonly ret_20d: Dec | null;
  readonly ret_90d: Dec | null;
  readonly ret_250d: Dec | null;
  readonly mkt_cap: Dec | null;
  readonly float_mkt_cap: Dec | null;
  readonly pe_ttm: Dec | null;
  readonly pe_dynamic: Dec | null;
  readonly pb: Dec | null;
  readonly peg: Dec | null;
  readonly gross_margin_ttm: Dec | null;
  /** Wave-quality composite ∈ [0, 1000]. Computed only by the batch
   *  backfill (needs universe-wide percentile tables — see
   *  `wcmi-scoring.ts` + `StockMetricsBackfillService`). Per-code
   *  projection leaves this field `null`; the next batch pass fills
   *  it in. */
  readonly wcmi: Dec | null;
  /** Per-sub-score cross-sectional percentile × 100. All `null` when
   *  `wcmi` is null. */
  readonly wcmi_rhythm: Dec | null;
  readonly wcmi_ma_support: Dec | null;
  readonly wcmi_up_wave: Dec | null;
  readonly wcmi_yang_dom: Dec | null;
  readonly wcmi_shadow_clean: Dec | null;
  readonly wcmi_stage_gain: Dec | null;
  readonly wcmi_crash_avoid: Dec | null;
  readonly wcmi_recent_strength: Dec | null;
}

const RETURN_WINDOWS: readonly (readonly [keyof StockMetrics, number])[] = [
  ['ret_1d', 1],
  ['ret_5d', 5],
  ['ret_10d', 10],
  ['ret_20d', 20],
  ['ret_90d', 90],
  ['ret_250d', 250],
];

const EMPTY_RETURNS: Pick<
  StockMetrics,
  'ret_1d' | 'ret_5d' | 'ret_10d' | 'ret_20d' | 'ret_90d' | 'ret_250d'
> = {
  ret_1d: null,
  ret_5d: null,
  ret_10d: null,
  ret_20d: null,
  ret_90d: null,
  ret_250d: null,
};

export function computeMetrics(meta: StockMetaDto, bars: readonly BarLike[]): StockMetrics {
  if (bars.length === 0) {
    return {
      code: meta.code,
      asof: null,
      price: null,
      ...EMPTY_RETURNS,
      mkt_cap: null,
      float_mkt_cap: null,
      pe_ttm: null,
      pe_dynamic: null,
      pb: null,
      peg: null,
      gross_margin_ttm: null,
      wcmi: null,
      wcmi_rhythm: null,
      wcmi_ma_support: null,
      wcmi_up_wave: null,
      wcmi_yang_dom: null,
      wcmi_shadow_clean: null,
      wcmi_stage_gain: null,
      wcmi_crash_avoid: null,
      wcmi_recent_strength: null,
    };
  }
  const latest = bars[bars.length - 1]!;
  const latestClose = new D(latest.close_qfq);
  const derived = deriveMetrics(meta, latestClose);
  const returns = computeReturns(bars, latestClose);
  return {
    code: meta.code,
    asof: latest.trade_date,
    price: latestClose.gt(0) ? latestClose : null,
    ...returns,
    ...derived,
    // wcmi is universe-rank-based — only the batch backfill knows
    // every code's raw features and can compute percentiles.
    wcmi: null,
    wcmi_rhythm: null,
    wcmi_ma_support: null,
    wcmi_up_wave: null,
    wcmi_yang_dom: null,
    wcmi_shadow_clean: null,
    wcmi_stage_gain: null,
    wcmi_crash_avoid: null,
    wcmi_recent_strength: null,
  };
}

function computeReturns(
  bars: readonly BarLike[],
  latestClose: Dec,
): Pick<StockMetrics, 'ret_1d' | 'ret_5d' | 'ret_10d' | 'ret_20d' | 'ret_90d' | 'ret_250d'> {
  if (latestClose.lte(0)) return EMPTY_RETURNS;
  const out: Record<string, Dec | null> = { ...EMPTY_RETURNS };
  for (const [name, window] of RETURN_WINDOWS) {
    if (bars.length <= window) {
      out[name] = null;
      continue;
    }
    const baseClose = new D(bars[bars.length - 1 - window]!.close_qfq);
    if (baseClose.lte(0)) {
      out[name] = null;
      continue;
    }
    out[name] = latestClose.sub(baseClose).div(baseClose);
  }
  return out as Pick<
    StockMetrics,
    'ret_1d' | 'ret_5d' | 'ret_10d' | 'ret_20d' | 'ret_90d' | 'ret_250d'
  >;
}
